src/main/java/uk/gov/justice/hmpps/datacompliance/client/duplicate/detection/AwsAthenaDuplicateDetectionClient.java

```diff
@@ -0,0 +1,117 @@
+package uk.gov.justice.hmpps.datacompliance.client.duplicate.detection;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+import software.amazon.awssdk.services.athena.AthenaClient;
+import software.amazon.awssdk.services.athena.model.GetQueryResultsResponse;
+import software.amazon.awssdk.services.athena.model.QueryExecutionState;
+import uk.gov.justice.hmpps.datacompliance.dto.OffenderNumber;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+import static java.lang.Math.pow;
+import static java.lang.String.format;
+import static java.lang.Thread.currentThread;
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.toSet;
+import static software.amazon.awssdk.services.athena.model.QueryExecutionState.QUEUED;
+import static software.amazon.awssdk.services.athena.model.QueryExecutionState.RUNNING;
+import static software.amazon.awssdk.services.athena.model.QueryExecutionState.SUCCEEDED;
+
+@Slf4j
+@Component
+@ConditionalOnProperty(name = "duplicate.detection.provider")
+public class AwsAthenaDuplicateDetectionClient implements DuplicateDetectionClient {
+
+    private static final Set<QueryExecutionState> ATHENA_WAITING_STATES = Set.of(QUEUED, RUNNING);
+    private static final int WAIT_TIME_MULTIPLIER = 2;
+    private static final int MAX_WAIT_ATTEMPTS = 8;
+
+    private final AthenaClient athenaClient;
+    private final DuplicateDetectionQueryFactory queryFactory;
+    private final Duration initialWaitDuration;
+
+    AwsAthenaDuplicateDetectionClient(final AthenaClient athenaClient,
+                                      final DuplicateDetectionQueryFactory queryFactory,
+                                      @Value("${duplicate.detection.athena.initial.wait:500ms}") final Duration initialWaitDuration) {
+
+        log.info("Configured to use AWS Athena for querying for offender duplicates");
+
+        this.athenaClient = athenaClient;
+        this.initialWaitDuration = initialWaitDuration;
+        this.queryFactory = queryFactory;
+    }
+
+    @Override
+    public Set<DuplicateResult> findDuplicatesFor(final OffenderNumber offenderNumber) {
+
+        final var queryExecutionId = submitAthenaQuery(offenderNumber);
+
+        awaitResponse(queryExecutionId);
```

Comment:
Good question.  So it will be the thread pool servicing the SQS queue.  I think this is ok because I do still want a deletion referral queue object to be placed in the dead letter queue if all of this goes wrong.

I didn't think that there were any async callback solutions, but I have just found:
https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/athena/AmazonAthenaAsync.html (search getQueryExecutionAsync) which might be what I'm looking for.

Reply:
So maybe it just doesn't matter, and it should be synchronous (else you can't retry by throwing an exception)